fw/http.c

```diff
@@ -5039,6 +5042,12 @@ tfw_http_req_cache_cb(TfwHttpMsg *msg)
 	/* Account current request in APM health monitoring statistics */
 	tfw_http_hm_srv_update((TfwServer *)srv_conn->peer, req);
 
```

Comment:
At this place, between getting the `srv_conn` and setting `TFW_CONN_B_UNSCHED`, we have a race: another CPU can get the same server connection and send an HTTP request to the same server socket right after our Upgrade request. Moreover, `tfw_http_popreq()` forwards pipelined requests, so the race is very probable. I spent quite a time to think how to fix the problem properly and didn't find an elegant solution.

It seem I also was wrong with my statement in the last review:
> Connection upgrade requests are non-idempotent by their nature. We can schedule (but not send!) HTTP requests to connections processing non-idempotent requests, but we should not schedule new requests for a connection processing upgrade message. This means that we can no just use TFW_CONN_B_HASNIP for upgrade requests and do need TFW_CONN_B_UNSCHED.

We schedule requests to `TfwSrvConn`, which remains intact during connection upgrade and only underlying socket is stolen. After revision it seems that we actually can treat connection upgrade requests as NIP and `tfw_http_conn_on_hold()` will do the work for us.

`tfw_http_conn_on_hold()` "unlocks" a connection, when a nip request is removed, i.e. by `tfw_http_propreq()` called from `tfw_http_resp_cache()`. `tfw_http_websocket_upgrade()` is called right after `tfw_http_resp_cache()`.

`tfw_http_resp_cache()` may send an error response to the client (by HTTP limits decision in the GFSM call), so the function must return error code an we should not upgrade to websockets on error.

It looks like the best way to fix the issue is to move `tfw_http_popreq()` our from `tfw_http_resp_cache()` and call it after `tfw_http_websocket_upgrade()`. Don't forget to call it also after `tfw_http_resp_cache()`  in `tfw_http_resp_terminate()`. We don't need TFW_CONN_B_UNSCHED in this case

Reply:
Found bug is legit, but to me nip request concept is unrelated here. Semantically nip requests cannot be pipelined, but idempotent requests _can be_ pipelined after it. In our case that is exactly the situation we want to avert. No one request must be sent after upgrade request forwarded to backend.

Overall your suggested fix can help, but to me it seams a little brittle. May be all races arise from a fact that we meddle with connection while changes to it drastically change its semantics. May be original approach with provisioning new connection instead of socket steal was not so wrong.

Note, that with your approach upgrade request in fact would block connection with backend until it respond to it. And what if `conns_n=1` in config?

You discard approach with provisioning new connection on a basis of locking server connection list argument. But what if we do not add it to the list or any other list, but give ownership over it to client connection. Specifically if we create separate connection outside the list only to forward upgrade request to it and may be upgrade it. It will be unavailable to any other request.

Reply:
BTW, for now i implemented your approach in last comment.

Reply:
> Semantically nip requests cannot be pipelined, but idempotent requests can be pipelined after it.
Actually vise versa: we do pipline a non-idempotent request after idempotent requests, but do not pipline after it.

> what if conns_n=1 in config?

Such configuration works, but I'd say this is 'misconfiguration'. Traditional servers like Nginx or HAproxy establish a new connection with backend for each concurrent request, so a real life Tempesta FW configuration must have way more connections, like 100 and more.

Reply:
I have not found a fix that at the same time was not invasive. So i implemented approach. I have leaved UNSCHED flag on place but introduced marking upgrade request as non-idempotent. This as i can see prevent race condition found by you.